socks/pool.go

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) 2019 The Decred developers
+// Use of this source code is governed by a 3-clause BSD
+// license that can be found in the LICENSE file.
+
+package socks
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync"
+)
+
+type Pool struct {
+	mtx     sync.Mutex
+	numOpen uint32
+	maxOpen uint32
+
+	proxy Proxy
+}
+
+func NewPool(proxy Proxy, maxOpen uint32) *Pool {
+	pool := Pool{
+		maxOpen: maxOpen,
+		proxy:   proxy,
+	}
+
+	return &pool
+}
+
+func (p *Pool) DialContext(ctx context.Context, network, address string) (net.Conn, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+	}
+
+	p.mtx.Lock()
+	maxOpen := p.maxOpen
+	if p.numOpen+1 > maxOpen {
+		p.mtx.Unlock()
+		return nil, fmt.Errorf("pool connection limit '%d' reached", maxOpen)
```

Comment:
can you return or wrap a specific error here? It will make it possible to detect these hit limits in dcrwallet.